jerry-core/ecma/builtin-objects/ecma-builtin-global.cpp

```diff
@@ -1362,6 +1362,103 @@ ecma_builtin_global_object_escape (ecma_value_t this_arg __attr_unused___, /**<
   return ret_value;
 } /* ecma_builtin_global_object_escape */
 
+/**
+ * The Global object's 'unescape' routine
+ *
+ * See also:
+ *          ECMA-262 v5, B.2.2
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_global_object_unescape (ecma_value_t this_arg __attr_unused___, /**< this argument */
+                                     ecma_value_t arg) /**< routine's first argument */
+{
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+
+  /* 1. */
+  ECMA_TRY_CATCH (string, ecma_op_to_string (arg), ret_value);
+  ecma_string_t *input_string_p = ecma_get_string_from_value (string);
+  /* 2. */
+  lit_utf8_size_t input_size = ecma_string_get_size (input_string_p);
+
+  /* 3. */
+  MEM_DEFINE_LOCAL_ARRAY (input_start_p, input_size, lit_utf8_byte_t);
+  ssize_t sz = ecma_string_to_utf8_string (input_string_p, input_start_p, (ssize_t) (input_size));
+  JERRY_ASSERT (sz >= 0);
+
+  lit_utf8_byte_t *input_curr_p = input_start_p;
+  lit_utf8_byte_t *input_end_p = input_start_p + input_size;
+  /* 4. */
+  /* The length of input string is always greater than output string
+   * so we re-use the input string buffer.
+   * The %xx is three byte long, and the maximum encoded value is 0xff,
+   * which maximum encoded length is two byte. Similar to this, the maximum
+   * encoded length of %uxxxx is four byte. */
+  lit_utf8_byte_t *output_char_p = input_start_p;
+
+  /* The state of parsing that tells us where we are in an escape pattern.
+   * 0    we are outside of pattern,
+   * 1    found '%', start of pattern,
+   * 2    found first hex digit of '%xy' pattern
+   * 3    found valid '%xy' pattern
+   * 4    found 'u', start of '%uwxyz' pattern
+   * 5-7  found hex digits of '%uwxyz' pattern
+   * 8    found valid '%uwxyz' pattern
+   */
+  uint8_t status = 0;
+  lit_code_point_t hex_digits = 0;
```

Comment:
It's enough to use an `ecma_char_t` for `hex_digits`, the decoded value will always be less than 0xFFFF.